types/underscore/underscore-tests.ts

```diff
@@ -887,6 +893,154 @@ declare const extractChainTypes: ChainTypeExtractor;
+    _(stringRecordDictionary).inject(stringRecordDictionaryMemoIterator, stringMemo); // $ExpectType string
+    _(stringRecordDictionary).inject(stringRecordPartialMemoIterator, stringMemo, context); // $ExpectType string
+    extractChainTypes(_.chain(stringRecordDictionary).inject(stringRecordDictionaryMemoIterator, stringMemo)); // $ExpectType ChainType<string, string>
+    extractChainTypes(_.chain(stringRecordDictionary).inject(stringRecordPartialMemoIterator, stringMemo, context)); // $ExpectType ChainType<string, string>
+
+    // with object memo - strings
+    _.reduce(simpleString, stringListMemoIterator, dictionaryMemo); // $ExpectType Dictionary<number>
+    _.reduce(simpleString, stringListMemoIterator, dictionaryMemo, context); // $ExpectType Dictionary<number>
+    _(simpleString).reduce(stringListMemoIterator, dictionaryMemo); // $ExpectType Dictionary<number>
+    _(simpleString).reduce(stringListMemoIterator, dictionaryMemo, context); // $ExpectType Dictionary<number>
+    extractChainTypes(_.chain(simpleString).reduce(stringListMemoIterator, dictionaryMemo)); // $ExpectType ChainType<Dictionary<number>, number>
+    extractChainTypes(_.chain(simpleString).reduce(stringListMemoIterator, dictionaryMemo, context)); // $ExpectType ChainType<Dictionary<number>, number>
+    _.foldl(simpleString, stringListMemoIterator, dictionaryMemo); // $ExpectType Dictionary<number>
+    _.foldl(simpleString, stringListMemoIterator, dictionaryMemo, context); // $ExpectType Dictionary<number>
+    _(simpleString).foldl(stringListMemoIterator, dictionaryMemo); // $ExpectType Dictionary<number>
+    _(simpleString).foldl(stringListMemoIterator, dictionaryMemo, context); // $ExpectType Dictionary<number>
+    extractChainTypes(_.chain(simpleString).foldl(stringListMemoIterator, dictionaryMemo)); // $ExpectType ChainType<Dictionary<number>, number>
+    extractChainTypes(_.chain(simpleString).foldl(stringListMemoIterator, dictionaryMemo, context)); // $ExpectType ChainType<Dictionary<number>, number>
+    _.inject(simpleString, stringListMemoIterator, dictionaryMemo); // $ExpectType Dictionary<number>
+    _.inject(simpleString, stringListMemoIterator, dictionaryMemo, context); // $ExpectType Dictionary<number>
+    _(simpleString).inject(stringListMemoIterator, dictionaryMemo); // $ExpectType Dictionary<number>
+    _(simpleString).inject(stringListMemoIterator, dictionaryMemo, context); // $ExpectType Dictionary<number>
+    extractChainTypes(_.chain(simpleString).inject(stringListMemoIterator, dictionaryMemo)); // $ExpectType ChainType<Dictionary<number>, number>
+    extractChainTypes(_.chain(simpleString).inject(stringListMemoIterator, dictionaryMemo, context)); // $ExpectType ChainType<Dictionary<number>, number>
+
+    // without memo - strings
+    _.reduce(simpleString, stringListSelfMemoIterator); // $ExpectType string
+    _.reduce(simpleString, stringListSelfMemoIterator, undefined, context); // $ExpectType string
```

Comment:
Following the actual Underscore semantics and assuming `strictNullChecks`, this would actually `$ExpectType never`. Though I think you just shouldn't test this scenario.

Reply:
I disagree. An undefined memo uses the overload I added which originally forced a `TResult` of the collection element type exactly but with the changes that came about from your comment [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45893#discussion_r449912866) now defaults to the collection element type as the result type but allows any type union and forcibly adds the collection type to the type union to handle the single-item collection case.

I think the case of an undefined memo is an interesting and distinct scenario because Underscore skips calling iteratee on the first iteration, which has unique side effects from a results perspective.

Reply:
> I think the case of an undefined memo is an interesting and distinct scenario because Underscore skips calling iteratee on the first iteration, which has unique side effects from a results perspective.

No, it never does that if you explicitly pass `undefined` as the memo, only if you pass fewer than three arguments, as I previously pointed out [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45893#discussion_r449908468). See the source for yourself:

https://github.com/jashkenas/underscore/blob/e6af0f9ec7253392047d36a432c41288ed67f7ee/modules/_createReduce.js#L25

https://github.com/jashkenas/underscore/blob/e6af0f9ec7253392047d36a432c41288ed67f7ee/modules/_createReduce.js#L13-L16

Reply:
Ah, I see, I saw the `arguments.length < 3` in [this comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45893#discussion_r449908468) but didn't initially get what that meant. I understand now and will revise the signature of the first overload to compensate.